File: 3286. Find a Safe Walk Through a Grid.cpp
#define pi pair<int, int>
#define pii pair<int, pi>
class Solution {
public:
bool findSafeWalk(vector<vector<int>>& grid, int health) 
{
    int n = grid.size(), m = grid[0].size();
    vector<vector<int>> visited(n, vector<int>(m));

    vector<vector<int>> distance(n, vector<int>(m, INT_MAX));
    distance[0][0] = grid[0][0];

    priority_queue<pii, vector<pii>, greater<pii>> minHeap;
    minHeap.push({distance[0][0], {0, 0}});

    vector<pi> dir = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    while(not minHeap.empty())
    {
        auto [cost, idx] = minHeap.top(); minHeap.pop();
        auto [i, j] = idx;
        if(distance[i][j] < cost) continue;

        for(auto [ii, jj] : dir)
        {
            int new_i = i + ii, new_j = j + jj;
            if(new_i < 0 or new_j < 0 or new_i == n or new_j == m) 
                continue;
            if(grid[new_i][new_j] + cost < distance[new_i][new_j])
            {
                distance[new_i][new_j] = grid[new_i][new_j] + cost;
                minHeap.push({distance[new_i][new_j], {new_i, new_j}});
            }
        }
        if(i == n - 1 and j == m - 1) break;
    }
    
    return distance[n - 1][m - 1] < health;
}
};

File: 3286. Find a Safe Walk Through a Grid.py
class Solution:
    def findSafeWalk(self, grid: List[List[int]], health: int) -> bool:
        max_heap = []
        visited = set()
        heappush(max_heap, (-(health - grid[0][0]), (0 , 0)))
        dir = ((0, 1), (0, -1), (1, 0), (-1, 0))

        n, m = len(grid), len(grid[0])
        while max_heap:
            h, (i, j) = heappop(max_heap)
            h = -h
            if not h:
                return False
                
            if i == n - 1 and j == m - 1:
                return True

            for ii, jj in dir:
                ni, nj = ii + i, jj + j
                key = (ni, nj)
                if ni < 0 or nj < 0 or ni == n or nj == m or key in visited:
                    continue

                visited.add(key)
                heappush(max_heap, (-(h - grid[ni][nj]), (ni, nj)))
            
        return False